cpp/src/slice2cs/Gen.cpp

```diff
@@ -1946,19 +1937,14 @@ Slice::Gen::TypesVisitor::visitExceptionEnd(const ExceptionPtr& p)
         paramDecl.push_back(memberType + " " + memberName);
 
         // Look for non-nullable fields to be initialized by the secondary constructor.
-        if (!q->optional())
+        if (!q->optional() && isNonNullableReferenceType(q->type(), false))
         {
-            StructPtr st = dynamic_pointer_cast<Struct>(q->type());
-            if (dynamic_pointer_cast<Sequence>(q->type()) || dynamic_pointer_cast<Dictionary>(q->type()) ||
-                (st && isMappedToClass(st)))
+            secondaryCtorParams.push_back(memberType + " " + memberName);
+            if (find(dataMembers.begin(), dataMembers.end(), q) != dataMembers.end())
             {
-                secondaryCtorParams.push_back(memberType + " " + memberName);
-                if (find(dataMembers.begin(), dataMembers.end(), q) != dataMembers.end())
-                {
-                    secondaryCtorMemberNames.push_back(memberName);
-                }
-                hideParameterlessCtor = true;
+                secondaryCtorMemberNames.push_back(memberName);
             }
+            hideParameterlessCtor = true;
```

Comment:
What is the logic for hiding the parameterless constructor? The parameterless construct is only visible when it initializes all fields to correct values. Still, as soon as we have a non-nullable reference type we hide it because otherwise that member would be incorrectly initialized.

Reply:
I don't understand your question.

We generate a public parameterless constructor for mapped Slice classes and exceptions no matter what for unmarshaling. 

Then, there are two possibities:
- all the fields are initialized to reasonable values by this constructor (no `null!`) => it's visible and usable by the application
- some fields are non-nullable reference fields and initialized to `null!` => we hide it from the application as the application should not use this constructor. These `null!` are temporary since the Slice engine is about to unmarshal non-null values into these fields.